File: frontend/src/engine/capabilities/bulkFormula.ts
/**
 * bulkFormula – Write a formula to an entire column, stopping at the last data row.
 *
 * The formula template should reference the first data row (e.g. "=A2*B2").
 * The executor auto-increments row numbers for each subsequent row.
 *
 * Example:
 *   formula:     "=A2*B2"
 *   dataRange:   "A:B"          ← used to detect last row
 *   outputRange: "C:C"          ← where formulas are written
 *   hasHeaders:  true           ← skip row 1
 */

import { CapabilityMeta, BulkFormulaParams, StepResult, ExecutionOptions } from "../types";
import { registry } from "../capabilityRegistry";
import { resolveRange, resolveSheet, stripWorkbookQualifier } from "./rangeUtils";

const meta: CapabilityMeta = {
  action: "bulkFormula",
  description: "Write a formula to an entire column up to the last data row",
  mutates: true,
  affectsFormatting: false,
};

async function handler(
  context: Excel.RequestContext,
  params: BulkFormulaParams,
  options: ExecutionOptions,
): Promise<StepResult> {
  const { formula, outputRange, dataRange, hasHeaders = true } = params;

  if (options.dryRun) {
    return { stepId: "", status: "success", message: `Would write formula "${formula}" to ${outputRange} based on data in ${dataRange}` };
  }

  options.onProgress?.("Detecting last data row...");

  // Get the last used row from dataRange
  const ws = resolveSheet(context, dataRange);
  const dataRng = resolveRange(context, dataRange);
  let lastRow = 1;
  try {
    const used = dataRng.getUsedRange(false);
    used.load("rowCount, address");
    await context.sync();
    const cellPart = used.address.includes("!") ? used.address.split("!").pop()! : used.address;
    const m = cellPart.replace(/\$/g, "").match(/[A-Z]+(\d+)/);
    const startRow = m ? parseInt(m[1], 10) : 1;
    lastRow = startRow + used.rowCount - 1;
  } catch {
    ws.getUsedRange(false).load("rowCount");
    await context.sync();
    lastRow = ws.getUsedRange(false).rowCount;
  }

  const firstDataRow = hasHeaders ? 2 : 1;
  if (lastRow < firstDataRow) {
    return { stepId: "", status: "success", message: "No data rows found." };
  }
  const rowCount = lastRow - firstDataRow + 1;
  options.onProgress?.(`Writing formula to ${rowCount} rows...`);

  // Determine output column and start row
  const stripped = stripWorkbookQualifier(outputRange);
  const ref = stripped.includes("!") ? stripped.split("!")[1] : stripped;
  const outCol = ref.match(/[A-Z]+/)?.[0] ?? "C";

  // Build formula array, incrementing row numbers from the template
  // Template: "=A2*B2" → for row 3: "=A3*B3"
  const templateRow = formula.match(/\d+/)
    ? parseInt(formula.match(/\d+/)![0], 10)
    : firstDataRow;

  const formulas: string[][] = [];
  for (let r = firstDataRow; r <= lastRow; r++) {
    const offset = r - templateRow;
    const f = offset === 0
      ? formula
      : formula.replace(/([A-Z]+)(\d+)/g, (_, col, row) => `${col}${parseInt(row, 10) + offset}`);
    formulas.push([f]);
  }

  const outRng = ws.getRange(`${outCol}${firstDataRow}:${outCol}${lastRow}`);
  outRng.formulas = formulas;
  await context.sync();

  return {
    stepId: "",
    status: "success",
    message: `Wrote "${formula}" (adjusted) to ${rowCount} rows in column ${outCol}`,
  };
}

registry.register(meta, handler as any);
export { meta };

File: backend/app/execution/capabilities/bulk_formula.py
"""
bulkFormula — write a row-adjusted formula to a column up to the last data row.

Detects the last row of the data range (via xlwings' used-range resolution),
then replicates the template formula down the output column with row
references auto-incremented — exactly mirroring the TS implementation.

Template: "=A2*B2" at row 3 becomes "=A3*B3".
"""

from __future__ import annotations

import re
from typing import Any

from app.execution.base import ExecutorContext
from app.execution.capability_registry import registry
from app.execution.range_utils import parse_address, resolve_range


# Column-then-row references anywhere in the formula, e.g. A2, $B$17, AA100.
# We need to rewrite the row half only; leaving any `$` prefix intact.
_CELL_REF_RE = re.compile(r"([A-Z]+)(\d+)")


def _first_row_number(formula: str) -> int:
    """Return the row number of the first cell reference in the formula."""
    m = re.search(r"\d+", formula)
    return int(m.group(0)) if m else 1


def _strip_workbook_qualifier(address: str) -> str:
    """Remove `[Book.xlsx]` qualifier for pattern-based parsing."""
    return re.sub(r"\[[^\]]+\]", "", address)


def handler(ctx: ExecutorContext, params: dict[str, Any]) -> dict[str, Any]:
    formula = params.get("formula")
    output_range = params.get("outputRange")
    data_range = params.get("dataRange")
    has_headers = params.get("hasHeaders", True)

    if not formula or not output_range or not data_range:
        return {
            "status": "error",
            "message": "bulkFormula requires 'formula', 'outputRange', and 'dataRange'.",
        }

    if ctx.dry_run:
        return {
            "status": "preview",
            "message": f"Would write formula {formula!r} to {output_range} based on data in {data_range}.",
        }

    try:
        data_rng = resolve_range(ctx.workbook_handle, data_range)
        sheet = data_rng.sheet

        # Determine the last row of actual data. Mirrors TS getUsedRange(false).
        try:
            used = data_rng.current_region if hasattr(data_rng, "current_region") else None
            if used is not None and used.count > 0:
                used_rows = used.shape[0]
                # Find the top row of data_rng (1-based).
                top_row = data_rng.row
                last_row = top_row + used_rows - 1
            else:
                raise AttributeError  # fall into fallback
        except Exception:  # noqa: BLE001 — fall back to whole-sheet used range
            whole = sheet.used_range
            last_row = whole.shape[0] if whole.shape else 1

        first_data_row = 2 if has_headers else 1
        if last_row < first_data_row:
            return {
                "status": "success",
                "message": "No data rows found.",
                "outputs": {},
            }

        row_count = last_row - first_data_row + 1

        # Determine output column letter.
        parsed_out = parse_address(output_range)
        out_cell = parsed_out.cell or "C1"
        out_cell_stripped = _strip_workbook_qualifier(out_cell)
        col_match = re.match(r"[A-Z]+", out_cell_stripped)
        out_col = col_match.group(0) if col_match else "C"

        # Build per-row formulas by offsetting every row-number digit.
        template_row = _first_row_number(formula)
        formulas: list[list[str]] = []
        for r in range(first_data_row, last_row + 1):
            offset = r - template_row
            if offset == 0:
                f = formula
            else:
                def _shift(m: re.Match[str]) -> str:
                    return f"{m.group(1)}{int(m.group(2)) + offset}"

                f = _CELL_REF_RE.sub(_shift, formula)
            formulas.append([f])

        out_addr = f"{out_col}{first_data_row}:{out_col}{last_row}"
        out_rng = sheet.range(out_addr)
        out_rng.formula = formulas
    except Exception as exc:  # noqa: BLE001
        return {
            "status": "error",
            "message": f"Bulk formula failed: {exc}",
            "error": str(exc),
        }

    return {
        "status": "success",
        "message": f"Wrote {formula!r} (adjusted) to {row_count} rows in column {out_col}.",
        "outputs": {"range": f"{out_col}{first_data_row}:{out_col}{last_row}"},
    }


registry.register("bulkFormula", handler, mutates=True)
